File: frontend/src/lib/plans.ts
// Single source of truth for InterviewX plan display/limits on the frontend.
// Mirrors backend/apps/subscriptions/plans.py — keep in sync.

export type PlanId = "free" | "pro" | "premium" | "max";

export interface PlanInfo {
  id: PlanId;
  label: string;
  priceRupees: number; // 0 for free
  monthlyLimit: number | null; // null = unlimited
  hasInsights: boolean; // dashboard history + AI topic insights
  features: string[];
}

export const PLANS: Record<PlanId, PlanInfo> = {
  free: {
    id: "free",
    label: "Free",
    priceRupees: 0,
    monthlyLimit: 2,
    hasInsights: false,
    features: ["2 AI mock interviews / month", "Limited companies & roles"],
  },
  pro: {
    id: "pro",
    label: "Pro",
    priceRupees: 199,
    monthlyLimit: 20,
    hasInsights: true,
    features: [
      "20 AI mock interviews / month",
      "All companies, roles & rounds",
      "Full dashboard & interview history",
      "AI insights: topic-by-topic scoring",
    ],
  },
  premium: {
    id: "premium",
    label: "Premium",
    priceRupees: 299,
    monthlyLimit: 50,
    hasInsights: true,
    features: [
      "50 AI mock interviews / month",
      "All companies, roles & rounds",
      "Full dashboard & interview history",
      "AI insights: topic-by-topic scoring",
      "Priority support",
    ],
  },
  max: {
    id: "max",
    label: "Max",
    priceRupees: 599,
    monthlyLimit: null,
    hasInsights: true,
    features: [
      "Unlimited AI mock interviews",
      "All companies, roles & rounds",
      "Full dashboard & interview history",
      "AI insights: topic-by-topic scoring",
      "Priority support",
    ],
  },
};

export const PAID_PLAN_IDS: PlanId[] = ["pro", "premium", "max"];

export function planOf(subscriptionPlan: string | undefined | null): PlanInfo {
  return PLANS[(subscriptionPlan as PlanId) || "free"] ?? PLANS.free;
}

export function isPaidPlan(subscriptionPlan: string | undefined | null): boolean {
  return PAID_PLAN_IDS.includes((subscriptionPlan as PlanId) || "free");
}

// Skills section (skill-only practice interviews) is Premium & Max only.
const SKILLS_PLAN_IDS: PlanId[] = ["premium", "max"];

export function hasSkills(subscriptionPlan: string | undefined | null): boolean {
  return SKILLS_PLAN_IDS.includes((subscriptionPlan as PlanId) || "free");
}

// ── Top-up packs ─────────────────────────────────────────────────────────────
// One-off interview credit packs, purchasable mid-cycle on top of the
// monthly plan limit. Mirrors backend TOPUP_PACKS — keep in sync.

export type TopupPackId = "spark" | "boost" | "power";

export interface TopupPackInfo {
  id: TopupPackId;
  label: string;
  priceRupees: number;
  credits: number;
  perInterview: string; // display string, e.g. "₹19.8/interview"
}

export const TOPUP_PACKS: Record<TopupPackId, TopupPackInfo> = {
  spark: { id: "spark", label: "Spark Pack", priceRupees: 99, credits: 5, perInterview: "₹19.8/interview" },
  boost: { id: "boost", label: "Boost Pack", priceRupees: 249, credits: 15, perInterview: "₹16.6/interview" },
  power: { id: "power", label: "Power Pack", priceRupees: 449, credits: 30, perInterview: "₹15.0/interview" },
};

export const TOPUP_PACK_IDS: TopupPackId[] = ["spark", "boost", "power"];

File: backend/apps/subscriptions/plans.py
"""
Single source of truth for InterviewX subscription plans.
Used by both the interviews app (limit checks) and subscriptions app (payments).
"""

# None = unlimited
PLANS = {
    "free": {
        "label": "Free",
        "amount_paise": 0,
        "monthly_limit": 2,
        "has_insights": False,
    },
    "pro": {
        "label": "Pro",
        "amount_paise": 19900,  # ₹199
        "monthly_limit": 20,
        "has_insights": True,
    },
    "premium": {
        "label": "Premium",
        "amount_paise": 29900,  # ₹299
        "monthly_limit": 50,
        "has_insights": True,
    },
    "max": {
        "label": "Max",
        "amount_paise": 59900,  # ₹599
        "monthly_limit": None,  # unlimited
        "has_insights": True,
    },
}

PAID_PLANS = {k: v for k, v in PLANS.items() if k != "free"}


# ---------------------------------------------------------------------------
# Mid-month top-up packs
# ---------------------------------------------------------------------------
# One-off interview credits a user can buy any time to go past their plan's
# monthly limit without waiting for the next billing cycle. Priced so that
# upgrading a tier remains the better deal per-interview (top-ups are always
# more expensive per interview than the next plan up).
TOPUP_PACKS = {
    "spark": {
        "label": "Spark Pack",
        "amount_paise": 9900,   # ₹99
        "credits": 5,           # ₹19.8 / interview
    },
    "boost": {
        "label": "Boost Pack",
        "amount_paise": 24900,  # ₹249
        "credits": 15,          # ₹16.6 / interview
    },
    "power": {
        "label": "Power Pack",
        "amount_paise": 44900,  # ₹449
        "credits": 30,          # ₹15.0 / interview
    },
}


def topup_amount_for(pack: str):
    return TOPUP_PACKS[pack]["amount_paise"]


def topup_credits_for(pack: str):
    return TOPUP_PACKS[pack]["credits"]


def monthly_limit_for(plan: str):
    return PLANS.get(plan, PLANS["free"])["monthly_limit"]


def amount_for(plan: str):
    return PLANS[plan]["amount_paise"]


def has_insights(plan: str) -> bool:
    """Only paid plans (Pro/Premium/Max) get topic-level AI insights &
    a full history dashboard. Free plan gets a plain overall score."""
    return PLANS.get(plan, PLANS["free"])["has_insights"]


# Plans that unlock the Skills section (practice interviews scoped to a
# single skill rather than a real company). Premium and Max only.
SKILLS_PLANS = {"premium", "max"}


def has_skills(plan: str) -> bool:
    return plan in SKILLS_PLANS
